requests-for-comments/2024-01-23-project-environments.md

```diff
@@ -147,42 +153,39 @@ Of course, nothing comes for free: adding evironments has its own downsides. Her
 
 ### Pricing
 
-Environments should start out as a paid feature, since they're key for professional users, and more of a nice-to-have for hobbyists. I would suggest it falling under the same restrictions as multiple projects, i.e. multiple environments being available after putting in a card.
+Having multiple environments should start out as a Teams/Enterprise feature.
 
 ### Rough scope of work
 
-1. Backend: We add an `Environment` model (`posthog_environment` table, fields `id`, `team`, `data_team_id`, `name`, `api_token`) and hook it up to a new project-scoped viewset (`/api/projects/:id/environments/`). Every time a new environment is created, we increment the counter of the `Team.id` field and use the obtained value as the environment's `data_team_id`.
-1. Backend: For every existing project (`Team`) we create a matching default `Environment`, reusing the project's `id` (for `data_team_id`) and `api_token` (for `api_token`). Any time a new project is created, it also gets a default environment based on the same logic.
-
-  > 	A transition period begins at this step. For some time we'll have to ensure `api_token` stays synced between the `Team` (legacy scheme) and its default `Environment` (future scheme). This only means updating the token reset function to update both.
-
+1. Backend: We add an `Environment` model (`posthog_environment` table, columns int `id`, `team_id`, `data_id` `name`, `api_token`) and hook it up to a new project-scoped viewset (`/api/projects/:id/environments/`). Every time a new environment is created, we determine its `id` by incrementing the counter of the `posthog_team.id` column and using that value as the environment's `data_id`.
+1. Backend: For every project existing currently we create a matching default `Environment`, reusing the project's `id` the environment's `data_id`. Any time a new project is created, it also gets a default environment based on the same logic.
+1. Backend: We add an `SDKKey` model (columns UUID `id`, `environment_id`, `value`, `created_by`, `created_at`) and populate it with current `api_token` values from `posthog_team` (for each project using the project ID as `environment_id`).
+    > Until we replace the project API key in project settings with environments, we'll have to sync the project API key being reset with its `posthog_sdkkey` equivalent.
```

Comment:
You don't need to create this table now, to avoid the syncing problem. We can for now keep using the `api_token` in the `posthog.team` table under the `data_id` team & not change anything in ingestion. It's still good to document the plan here though.

Reply:
We don't need to create it now, but we do probably need to create it early. Either we move to `SourceKey.value` or to `Environment.api_token`, but we definitely can't keep using `Team.api_token`.

Reply:
I'm not sure I follow where you're going here. My understanding is that events and persons are segmented per environment, so each environment needs its separate capture token / api_key. Even if we don't ship token-per-source yet, we'll need several tokens per project, hence the need for the separate table.

Similarly, plugin-server will need to resolve this per-environment token into an env and team for processing, so there's some changes to be done there.

After that table exists, shipping several tokens per env will be a frontend-only change.